File: Haskell/phoneBook.hs
import Data.Maybe

-- Define data type.
type Name = String
type PhoneNumber = String
type PhoneBook = [(Name,PhoneNumber)]

-- Searches for a phone number in a phone book.
-- Params: a name and a phone book.
-- Returns: the phone number if found, Nothing otherwise.
getPhoneNumber :: Name -> PhoneBook -> Maybe PhoneNumber
getPhoneNumber name [] = Nothing
getPhoneNumber name (x:xs)
    | name == (fst x) = Just (snd x)
    | otherwise = getPhoneNumber name xs

-- Searches for a phone number in a phone book, implemented with fold.
-- Params: a name and a phone book.
-- Returns: the phone number if found, Nothing otherwise.
getPhoneNumberFold :: Name -> PhoneBook -> Maybe PhoneNumber
getPhoneNumberFold name x = foldl (\ acc (k, v) -> if k == name then Just v else acc) Nothing x

File: Rust/phone_book.rs
fn get_phone_number(name: String, phone_book: [(String, i32); 3]) -> Result<i32, String> {

    let iterator = phone_book.iter();
    
    iterator.fold(Err(String::from("errore")), |acc, (k, v)| if *k == name {Ok(*v)} else {acc})
}

fn main() {
    let phone_book: [(String, i32); 3] = [(String::from("Pippo"), 123456), (String::from("Minnie"), 987654), (String::from("Pluto"), 456789)];
    let result = get_phone_number(String::from("Pippo"), phone_book);

    if result.is_ok() {
        println!("{}", result.unwrap())
    }
    else {
        println!("Not found");
    }
}